java/src/main/java/com/logicalclocks/hsfs/constructor/Query.java

```diff
@@ -41,39 +45,33 @@
 import java.util.stream.Collectors;
 
 @NoArgsConstructor
-public class Query {
-
-  private static final Logger LOGGER = LoggerFactory.getLogger(FeatureGroup.class);
-
-  @Getter
-  @Setter
-  private FeatureGroupBase leftFeatureGroup;
-  @Getter
-  @Setter
-  private List<Feature> leftFeatures;
-  @Getter
-  @Setter
-  private Long leftFeatureGroupStartTime;
-  @Getter
-  @Setter
-  private Long leftFeatureGroupEndTime;
-  @Getter
-  @Setter
-  private List<Join> joins = new ArrayList<>();
-  @Getter
-  @Setter
-  private FilterLogic filter;
-  @Getter
-  @Setter
-  private Boolean hiveEngine = false;
+public class Query extends QueryBase {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Query.class);
 
   private QueryConstructorApi queryConstructorApi = new QueryConstructorApi();
   private StorageConnectorApi storageConnectorApi = new StorageConnectorApi();
-  private FeatureGroupUtils utils = new FeatureGroupUtils();
 
   public Query(FeatureGroupBase leftFeatureGroup, List<Feature> leftFeatures) {
     this.leftFeatureGroup = leftFeatureGroup;
-    this.leftFeatures = leftFeatures;
+    this.leftFeatures = addFeatureGroupToFeatures(leftFeatureGroup, leftFeatures);
+  }
+
+  @Override
+  public String sql() {
+    // overriding toString does not work wtih jackson
+    return sql(Storage.OFFLINE);
+  }
+
+  @Override
+  public String sql(Storage storage) {
+    try {
+      return queryConstructorApi
+          .constructQuery(this.getLeftFeatureGroup().getFeatureStore(), this, FsQuery.class)
+          .getStorageQuery(storage);
+    } catch (FeatureStoreException | IOException e) {
+      return e.getMessage();
+    }
   }
 
   public Query join(Query subquery) {
```

Comment:
This and other methods are not defined in the base class. This could cause issues later on if we create a new class that extends the base class but we forget to make the method. I think all user-facing (aka. public) methods should be stored in the base class as abstract (applies to all classes).

Reply:
generally agree, but the problem is that I can't extend `QueryBase join(QueryBase subquery)` and it important that it returns `Query` object and not  `QueryBase`

Reply:
defining `public abstract QueryBase join(Query subquery);` in base class is enough, you can return it as `Query`. There are no issues as long as  `Query` extends  `QueryBase`.